addons/l10n_es_edi_verifactu/models/verifactu_document.py

```diff
@@ -0,0 +1,1186 @@
+
+            exempt_reason = tax_detail['l10n_es_exempt_reason']  # only set if exempt
+
+            tax_percentage = tax_detail['amount']
+            base_amount = sign * tax_detail['base_amount']
+            tax_amount = math.copysign(tax_detail['tax_amount'], base_amount)
+
+            calificacion_operacion = None  # Reported if not tax-exempt;
+            recargo_equivalencia = {}
+            if tax_type in sujeto_tax_types:
+                calificacion_operacion = 'S2' if tax_type == 'sujeto_isp' else 'S1'
+                if tax_detail['recargo_taxes']:
+                    recargo_key = recargo_tax_details_key.get(key)
+                    recargo_tax_detail = tax_details['tax_details'][recargo_key]
+                    recargo_tax_percentage = recargo_tax_detail['amount']
+                    recargo_tax_amount = math.copysign(recargo_tax_detail['tax_amount'], base_amount)
+                    recargo_equivalencia.update({
+                        'tax_percentage': recargo_tax_percentage,
+                        'tax_amount': recargo_tax_amount,
+                    })
+            elif tax_type in ('no_sujeto', 'no_sujeto_loc'):
+                calificacion_operacion = 'N2' if tax_type == 'no_sujeto_loc' else 'N1'
+            else:
+                # tax_type == 'exento' (see `_check_record_values`)
+                # exempt_reason set already
+                # [1238]
+                #     Si la operacion es exenta no se puede informar ninguno de los campos
+                #     TipoImpositivo, CuotaRepercutida, TipoRecargoEquivalencia y CuotaRecargoEquivalencia.
+                tax_percentage = None
+                tax_amount = None
+                recargo_percentage = None
+                recargo_amount = None
+
+            recargo_percentage = recargo_equivalencia.get('tax_percentage')
+            recargo_amount = recargo_equivalencia.get('tax_amount')
+
+            # Note on the TipoImpositivo and CuotaRepercutida tags.
+            # In some cases it makes a difference for the validation whether the tags are output with 0
+            # or not at all:
+            # - In the no sujeto cases (calification_operacion in ('N1', 'N2')) we may not include them.
+            # - In the (calification_operacion == S2) case the tags have to be included with value 0.
+            #
+            # See the following errors
+            # [1198]
+            #     Si CalificacionOperacion es S2 TipoImpositivo y CuotaRepercutida deberan tener valor 0.
+            if calificacion_operacion in ('N1', 'N2') and verifactu_tax_type == '01':
+                tax_percentage = None
+                tax_amount = None
+
+            detalle = {
+                'Impuesto': verifactu_tax_type,
+                'ClaveRegimen': clave_regimen,
+                'CalificacionOperacion': calificacion_operacion,
+                'OperacionExenta': exempt_reason,
+                'TipoImpositivo': self._format_number_Tipo2_2(tax_percentage),
```

Comment:
If it is 0 is the field filtered?  Did you test 0% taxes? (no_sujeto, no_sujeto_loc, ...)